File: src/news_digest/dates.py
"""Canonical date utilities for News Digest Agent — issue #102.

The app treats **America/New_York** as its calendar timezone: the ``digest_date``
column in Supabase reflects the Eastern date on which the digest was generated,
not the UTC date.  This matters most around midnight Eastern time when the UTC
and Eastern dates differ by one day.

Public API
----------
APP_TIMEZONE : str
    The IANA timezone name read from ``settings.app_timezone``.  Exposed for
    callers that need the string (e.g. APScheduler job registration).

app_today() -> datetime.date
    Return today's date in the configured app timezone.  This is the single
    source of truth for "what calendar date is it?" across the codebase.

One-time discontinuity note
---------------------------
Before issue #102, ``push_to_supabase`` stamped ``digest_date`` with
``datetime.now(UTC).date()``.  After this change it uses ``app_today()``
(Eastern).  The few rows written with a UTC date before the deploy will retain
their original dates and will be purged within ``retention_days`` (3 by default)
of the deploy date — no migration is needed.
"""

from datetime import UTC, date, datetime
from zoneinfo import ZoneInfo

from news_digest.config import get_settings


def _tz() -> ZoneInfo:
    """Return the app ZoneInfo object, reading timezone from settings each call.

    Not cached: settings-level caching already avoids repeated env reads.
    """
    return ZoneInfo(get_settings().app_timezone)


def app_today() -> date:
    """Return today's calendar date in the configured app timezone.

    Always use this instead of ``datetime.now(UTC).date()`` for any logic that
    answers "what calendar date is it right now?" — digest stamping, published-
    today checks, retention cutoff calculation, etc.

    Returns:
        The current date in ``settings.app_timezone`` (default
        ``America/New_York``).
    """

    now_utc = datetime.now(UTC)
    return now_utc.astimezone(_tz()).date()


#: Exposed for callers that need the IANA timezone name string (e.g. to
#: register APScheduler jobs in the app's local timezone).
APP_TIMEZONE: str = "America/New_York"  # overridden at runtime by settings

File: web/src/lib/dates.ts
// Date utilities for the News Digest app.
// All "today" comparisons use America/New_York — the agent runs on Eastern time,
// so that is the canonical date boundary for digest publication.

export const APP_TIMEZONE = "America/New_York";

/**
 * Return today's calendar date as a YYYY-MM-DD string in America/New_York.
 * Accepts an optional `now` for testability; defaults to the current instant.
 */
export function appToday(now: Date = new Date()): string {
  return new Intl.DateTimeFormat("en-CA", { timeZone: APP_TIMEZONE }).format(now);
}

/**
 * Return true if `digestDate` (YYYY-MM-DD) matches today in America/New_York.
 * Accepts an optional `now` for testability.
 */
export function isToday(digestDate: string, now: Date = new Date()): boolean {
  return digestDate === appToday(now);
}
